packages/flutter/lib/src/widgets/magnifier.dart

```diff
@@ -0,0 +1,545 @@
+  /// the magnifier to the shape. If null, the shape will be rectangular.
+  final ShapeBorder shape;
+
+  @override
+  RenderObject createRenderObject(BuildContext context) {
+    return _RenderMagnification(focalPoint, magnificationScale, shape);
+  }
+
+  @override
+  void updateRenderObject(
+      BuildContext context, covariant RenderProxyBox renderObject) {
+    (renderObject as _RenderMagnification)
+      ..focalPoint = focalPoint
+      ..shape = shape
+      ..magnificationScale = magnificationScale;
+  }
+}
+
+class _RenderMagnification extends RenderProxyBox {
+  _RenderMagnification(
+    this._focalPoint,
+    this._magnificationScale,
+    this._shape, {
+    RenderBox? child,
+  }) : super(child);
+
+  Offset get focalPoint => _focalPoint;
+  Offset _focalPoint;
+  set focalPoint(Offset value) {
+    if (_focalPoint == value) {
+      return;
+    }
+    _focalPoint = value;
+    markNeedsLayout();
+  }
+
+  double get magnificationScale => _magnificationScale;
+  double _magnificationScale;
+  set magnificationScale(double value) {
+    if (_magnificationScale == value) {
+      return;
+    }
+    _magnificationScale = value;
+    markNeedsLayout();
+  }
+
+  ShapeBorder get shape => _shape;
+  ShapeBorder _shape;
+  set shape(ShapeBorder value) {
+    if (_shape == value) {
+      return;
+    }
+    _shape = value;
+    markNeedsLayout();
+  }
+
+  @override
+  _MagnificationLayer? get layer => super.layer as _MagnificationLayer?;
+
+  @override
+  void paint(PaintingContext context, Offset offset) {
+    if (layer == null) {
+      layer = _MagnificationLayer(
+          size: size,
+          globalPosition: offset,
+          focalPoint: focalPoint,
+          shape: shape,
+          magnificationScale: magnificationScale);
+    } else {
+      layer!
+        ..magnificationScale = magnificationScale
+        ..size = size
+        ..globalPosition = offset
+        ..focalPoint = focalPoint;
+    }
+
+    context.pushLayer(layer!, super.paint, offset);
+  }
+}
+
+class _MagnificationLayer extends ContainerLayer {
+  _MagnificationLayer({
+      required this.size,
+      required this.globalPosition,
+      required this.shape,
+      required this.focalPoint,
+      required this.magnificationScale,
+  });
+
+  Offset globalPosition;
+  Size size;
+
+  Offset focalPoint;
+  double magnificationScale;
+
+  ShapeBorder shape;
+
+  @override
+  void addToScene(SceneBuilder builder) {
+    builder.pushClipPath(shape.getOuterPath(globalPosition & size));
+
+    final Offset thisCenter = Alignment.center.alongSize(size) + globalPosition;
+    final Matrix4 matrix = Matrix4.identity()
+      ..translate(
+          magnificationScale * (focalPoint.dx - thisCenter.dx) + thisCenter.dx,
+          magnificationScale * (focalPoint.dy - thisCenter.dy) + thisCenter.dy)
+      ..scale(magnificationScale);
+    builder.pushBackdropFilter(ImageFilter.matrix(matrix.storage, filterQuality: FilterQuality.high));
```

Comment:
Should we reuse the engine layer here?

Reply:
I'll have to look into how to do this, but this seems sensible

Reply:
By that I was referring to what the `BackdropFIlterLayer` does in its `addToScene`